gcc/rust/resolve/rust-forever-stack.hxx

```diff
@@ -365,6 +365,52 @@ check_leading_kw_at_start (const S &segment, bool condition)
   return condition;
 }
 
+#define unwrap_type_segment(x)                                                 \
+  (unwrap_type_segment_inner<typename std::remove_const<                       \
+     typename std::remove_reference<decltype (x)>::type>::type>::unwrap (x))
+
+template <class T> class unwrap_type_segment_inner;
+
+template <> class unwrap_type_segment_inner<AST::SimplePathSegment>
+{
+public:
+  using ret = AST::SimplePathSegment;
+
+  static AST::SimplePathSegment &unwrap (AST::SimplePathSegment &x)
+  {
+    return x;
+  }
+
+  static const AST::SimplePathSegment &unwrap (const AST::SimplePathSegment &x)
+  {
+    return x;
+  }
+};
+
+template <class T> class unwrap_type_segment_inner<std::unique_ptr<T>>
+{
+public:
+  using ret = typename unwrap_type_segment_inner<T>::ret;
+
+  static ret &unwrap (std::unique_ptr<T> &x)
+  {
+    return unwrap_type_segment (*x);
+  }
+  static const ret &unwrap (const std::unique_ptr<T> &x)
+  {
+    return unwrap_type_segment (*x);
+  }
+};
+
+template <class T> class unwrap_type_segment_inner
+{
+public:
+  using ret = AST::PathIdentSegment;
+
+  static ret &unwrap (T &x) { return x.get_ident_segment (); }
+  static const ret &unwrap (const T &x) { return x.get_ident_segment (); }
+};
```

Comment:
is this really needed? I can't really wrap my head around that code or what code it "replaces" if that makes sense. it seems a bit hard to maintain, but if there's a lot to gain from using it I'm happy with it.

Reply:
It's supposed to help handle different segment types being passed to `resolve_path`, by converting segment references down into `PathIdentSegment` or `SimplePathSegment` references. I suppose I could try to reduce it down to a couple of templated function overloads

Reply:
When I try to reduce it down to function overloads I keep getting linker errors, since some of the functions aren't templates, and everything's defined in an `.hxx` file (included in multiple places)

Reply:
I've added some commits to try to make things more clear